File: power-adapters-sample/src/main/java/com/nextfaze/poweradapters/sample/NewsItemView.java
package com.nextfaze.poweradapters.sample;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.RelativeLayout;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnLongClick;
import lombok.NonNull;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.Iterator;
import java.util.Set;

import static java.lang.String.format;

@Accessors(prefix = "m")
public final class NewsItemView extends RelativeLayout {

    private static final int CLICK_COUNT = 1;
    private static final int LONG_CLICK_COUNT = 5;

    @BindView(R.id.title)
    TextView mTitleView;

    @BindView(R.id.tags)
    TextView mTagsView;

    @Setter
    @Nullable
    private OnRemoveListener mOnRemoveListener;

    @Setter
    @Nullable
    private OnInsertListener mOnInsertBeforeListener;

    @Setter
    @Nullable
    private OnInsertListener mOnInsertAfterListener;

    public NewsItemView(Context context) {
        this(context, null);
    }

    public NewsItemView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public NewsItemView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        inflate(context, R.layout.news_item, this);
        ButterKnife.bind(this);
    }

    @OnClick(R.id.remove_button)
    void onRemoveClick() {
        if (mOnRemoveListener != null) {
            mOnRemoveListener.onRemove(CLICK_COUNT);
        }
    }

    @OnLongClick(R.id.remove_button)
    boolean onRemoveLongClick() {
        if (mOnRemoveListener != null) {
            mOnRemoveListener.onRemove(LONG_CLICK_COUNT);
        }
        return true;
    }

    @OnClick(R.id.insert_before_button)
    void onInsertBeforeClick() {
        if (mOnInsertBeforeListener != null) {
            mOnInsertBeforeListener.onInsert(CLICK_COUNT);
        }
    }

    @OnLongClick(R.id.insert_before_button)
    boolean onInsertBeforeLongClick() {
        if (mOnInsertBeforeListener != null) {
            mOnInsertBeforeListener.onInsert(LONG_CLICK_COUNT);
        }
        return true;
    }

    @OnClick(R.id.insert_after_button)
    void onInsertAfterClick() {
        if (mOnInsertAfterListener != null) {
            mOnInsertAfterListener.onInsert(CLICK_COUNT);
        }
    }

    @OnLongClick(R.id.insert_after_button)
    boolean onInsertAfterLongClick() {
        if (mOnInsertAfterListener != null) {
            mOnInsertAfterListener.onInsert(LONG_CLICK_COUNT);
        }
        return true;
    }

    public void setNewsItem(@NonNull NewsItem newsItem) {
        mTitleView.setText(format("%s (%s)", newsItem.getTitle(), newsItem.getType().toString().toLowerCase()));
    }

    public void setTags(@NonNull Set<String> tags) {
        CharSequence formatted = formatTags(tags);
        mTagsView.setText(formatted);
        mTagsView.setVisibility(formatted != null ? VISIBLE : GONE);
    }

    @Nullable
    private CharSequence formatTags(@NonNull Set<String> tags) {
        if (tags.isEmpty()) {
            return null;
        }
        StringBuilder b = new StringBuilder();
        for (Iterator<String> it = tags.iterator(); it.hasNext(); ) {
            String tag = it.next();
            b.append(tag);
            if (it.hasNext()) {
                b.append(", ");
            }
        }
        return b;
    }

    public interface OnInsertListener {
        void onInsert(int count);
    }

    public interface OnRemoveListener {
        void onRemove(int count);
    }
}

File: power-adapters-sample/src/main/java/com/nextfaze/poweradapters/sample/news/NewsItemView.kt
package com.nextfaze.poweradapters.sample.news

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.widget.RelativeLayout
import com.nextfaze.poweradapters.sample.R
import kotlinx.android.synthetic.main.news_item_view.view.*
import kotlin.properties.Delegates.observable

private const val CLICK_COUNT = 1
private const val LONG_CLICK_COUNT = 5

@SuppressLint("SetTextI18n")
class NewsItemView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : RelativeLayout(context, attrs, defStyleAttr) {
    
    var newsItem by observable<NewsItem?>(null) { _, _, newsItem ->
        if (newsItem != null) titleView.text = "${newsItem.title} (${newsItem.type.toString().toLowerCase()})"
    }
    
    var tags by observable(emptySet<String>()) { _, _, tags ->
        val formatted = formatTags(tags)
        tagsView.text = formatted
        tagsView.visibility = if (formatted != null) VISIBLE else GONE
    }
    
    var onRemoveListener: ((Int) -> Unit)? = null

    var onInsertBeforeListener: ((Int) -> Unit)? = null

    var onInsertAfterListener: ((Int) -> Unit)? = null

    init {
        inflate(context, R.layout.news_item_view, this)
        removeButton.setOnClickListener { onRemoveListener?.invoke(CLICK_COUNT) }
        removeButton.setOnLongClickListener { onRemoveListener?.invoke(LONG_CLICK_COUNT); true }
        insertBeforeButton.setOnClickListener { onInsertBeforeListener?.invoke(CLICK_COUNT) }
        insertBeforeButton.setOnLongClickListener { onInsertBeforeListener?.invoke(LONG_CLICK_COUNT); true }
        insertAfterButton.setOnClickListener { onInsertAfterListener?.invoke(CLICK_COUNT) }
        insertAfterButton.setOnLongClickListener { onInsertAfterListener?.invoke(LONG_CLICK_COUNT); true }
    }
}

private fun formatTags(tags: Set<String>) = if (tags.isEmpty()) null else tags.joinToString(", ")
